training/train.py

```diff
@@ -127,54 +116,46 @@ def preprocess_raw_data(raw_data_dict, lower_limit, upper_limit):
     return package_id_dict, manifest_id_dict
 
 
-def preprocess_data(data_list, lower_limit, upper_limit):
-    """Preprocess data."""
-    if len(data_list) == 2:
-        package_dict = (data_list[0]).get('package_list')
-        manifest_dict = (data_list[1]).get('manifest_list')
-        logger.info("Size of Package ID dictionary {} and Manifest ID dictionary are: {}".format(
-            len(package_dict), len(manifest_dict)))
-        return package_dict, manifest_dict
-    else:
-        raw_data = data_list[0]
-        package_dict, manifest_dict = preprocess_raw_data(raw_data, lower_limit, upper_limit)
-        logger.info("Size of Package ID dictionary {} and Manifest ID dictionary are: {}".format(
-            len(package_dict), len(manifest_dict)))
-        return package_dict, manifest_dict
-
-
-def make_user_item_df(manifest_dict, package_dict):
+def make_user_item_df(manifest_dict, package_dict, user_input_stacks):
     """Make user item dataframe."""
     user_item_list = []
+    set_input_stacks = {frozenset(x) for x in user_input_stacks}
     id_package_dict = {v: k for k, v in package_dict.items()}
     for k, v in manifest_dict.items():
         user_id = int(k)
+        is_user_input_stack = v in set_input_stacks
         for package in v:
             if package in id_package_dict:
                 item_id = package
                 user_item_list.append(
                     {
                         "UserId": user_id,
                         "ItemId": item_id,
-                        "Count": 1
+                        "Count": 1,
+                        "is_user_input_stack": is_user_input_stack
                     }
                 )
     return user_item_list
 
 
 def train_test_split(data_df):
     """Split for training and testing."""
-    data_df = data_df.sample(frac=1)
-    df_user = data_df.drop_duplicates(['UserId'])
-    data_df = data_df.sample(frac=1)
-    df_item = data_df.drop_duplicates(['ItemId'])
+    user_input_df = data_df.loc[data_df['is_user_input_stack']]
+    user_input_df = user_input_df.sample(frac=1)
+    df_user = user_input_df.drop_duplicates(['UserId'])
+    user_input_df = user_input_df.sample(frac=1)
+    df_item = user_input_df.drop_duplicates(['ItemId'])
     train_df = pd.concat([df_user, df_item]).drop_duplicates()
-    fraction = round(frac(data_df, train_df), 2)
+    fraction = round(frac(user_input_df, train_df), 2)
 
     if fraction < 0.80:
-        df_ = extra_df(fraction, data_df, train_df)
+        df_ = extra_df(fraction, user_input_df, train_df)
         train_df = pd.concat([train_df, df_])
-    test_df = pd.concat([data_df, train_df]).drop_duplicates(keep=False)
+    test_df = pd.concat([user_input_df, train_df]).drop_duplicates(keep=False)
+    test_df = test_df.drop(columns=['is_user_input_stack'])
+    data_df = data_df.loc[~data_df['is_user_input_stack']]
+    train_df = pd.concat([data_df, train_df])
```

Comment:
Help me understand if training DataFrame has reached 80% of given data, then why we are adding more data into it?

Reply:
It's only 80% of user input data. What about bigquery data? We need to add the BQ data as well and hence we are adding 100% of BQ data. Hence total data is now 180% of total data.

Reply:
Understood thanks :)